File: Algorithm/019_Remove_Nth_Node_From_End.py
"""Remove n-th node from end of the list.
Return head node
"""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """

        slow = head
        fast = head

        for i in range(n):
            fast = fast.next
            
        if fast == None:
            # In this case, the node we need to delete is the first node.
            head = slow.next
        else:
            # find out the node we want to delete (slow.next)
            while fast.next != None:
                fast = fast.next
                slow = slow.next
            
            slow.next = slow.next.next

        return head
    
    def remove_Nth_From_End_With_Dummy(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        if not head:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        fast = dummy
        slow = dummy

        for i in range(n + 1):
            fast = fast.next

        while fast != None:
            fast = fast.next
            slow = slow.next
        
        slow.next = slow.next.next

        return dummy.next

File: Algorithm/019_Remove_Nth_Node_From_End.cpp

 // Definition for singly-linked list.
 struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head)
        return nullptr;

        ListNode new_head(-1);
        new_head.next = head;

        ListNode *slow = &new_head, *fast = &new_head;

        for (int i = 0; i < n; i++)
            fast = fast->next;

        while (fast->next) 
        {
            fast = fast->next;
            slow = slow->next;
        }

        ListNode* to_de_deleted = slow->next;
        slow->next = slow->next->next;

        delete to_de_deleted;

        return new_head.next;
    }
};